### DISCUSSION TITLE
Error when converting to products

### DISCUSSION THREAD

**Discussion by ukraine-sdg** [open-sdg/open-sdg#1951] [Q&A] [answered]
Good day! Please help to correct the error when transferring data to products https://github.com/ukraine-sdg/data/pull/2112. Is it enough to fix the error to replace in the files: deploy-to-production.yml
deploy-to-staging.yml
test-pull-requests.yml version 3.7.1 on 3.7? And tell me what are the problems with the 11.6.1 indicator? Thank you!

**Comment by brockfanning** [comment, marked as answer]
After following up separately, we figured out that the root problem here was that the production site was not automatically rebuilding after the production data was updated. So a quick workaround was to manually perform a production release. In a nutshell, those steps were:

1. Make any change to the develop branch (such as changing the README)
2. Perform a production release (ie, merge the develop branch into the master branch)

The longer-term fix involved the Github Actions "workflow" files. This is a good opportunity to detail the current best practice for automating production releases. Previously, we had recommended using the [repository_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch) approach, which involved something like this in the data repository's deploy-to-production.yml file:

```
    - name: Rebuild live site
      run: |
        curl -X POST https://api.github.com/repos/my-organization/my-site-repo/dispatches \
        -H "Accept: application/vnd.github.everest-preview+json" \
        -H "Authorization: token ${{ secrets.token }}" \
        --data '{"event_type": "rebuild-live"}'
```

However, this approach has a major problem, in that it actually deploys the site repository's *develop* branch -- not its *master* branch. So instead, we now recommend using the [workflow_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) approach. This instead looks like this:

```
    - name: Rebuild live site
      run: |
        curl -L \
          -X POST \
          -H "Accept: application/vnd.github+json" \
          -H "Authorization: Bearer ${{ secrets.token }}"\
          -H "X-GitHub-Api-Version: 2022-11-28" \
          https://api.github.com/repos/my-organization/my-site-repo/actions/workflows/deploy-to-production.yml/dispatches \
          -d '{"ref":"master"}'
```

Importantly, this new approach also requires a change in the *site repository's* deploy-to-production.yml file:

Instead of this:

```
on:
  push:
    branches:
      - master
  repository_dispatch:
    types: rebuild-live
```

It should instead have this:

```
on:
  push:
    branches:
      - master
  workflow_dispatch:
```

(though that "workflow_dispatch" looks like something is missing, that's actually correct)

The following two pull-requests demonstrate the change from the old way to the new way:

* https://github.com/ukraine-sdg/site/pull/797
* https://github.com/ukraine-sdg/data/pull/2197

**Reply by christoph-LSN** [reply]
@brockfanning do I have to do this changes in my repos too? I haven't observed this error yet.

Site repo: https://github.com/christoph-LSN/MT_Site
Data repo: https://github.com/christoph-LSN/IM_Data

**Reply by brockfanning** [reply]
@christoph-LSN You don't have automated production site builds setup. In other words, when you do a production data build, it does not automatically perform a production site build. It is up to you to wait several minutes and then manually do a production site build (as you know).

It is not surprising that you don't have this set up, because we don't have this in our "starter" repos. We do have it setup in our [deploy-to-staging.yml](https://github.com/open-sdg/open-sdg-data-starter/blob/develop/.github/workflows/deploy-to-staging.yml) files (see the "Rebuild site" section), but not in our [deploy-to-production.yml](https://github.com/open-sdg/open-sdg-data-starter/blob/develop/.github/workflows/deploy-to-production.yml) files.

Some folks have set this up by adding the necessary lines to their "deploy-to-production.yml" files (both the site and data one). However, it turns out that the approach that we currently use in our deploy-to-staging.yml files is **not the correct approach** when it comes to production builds. What I detailed above (using "workflow_dispatch") is the correct approach.

I think we need to update our starter repos to use the correct approach. To make a long story short - you don't need to change anything. Right now you're manually doing production site builds. But if you want to set up automatic production site builds, I'd recommend you use the "workflow_dispatch" detailed here.
